File: sdks/dotnet/src/PrismDb/Value.cs
// The scalar value model and its tagged/untagged wire codec.
//
// Mirrors crates/prism-protocol/src/data.rs (Value) and the type tags in
// docs/specs/record-format.md. The SDK accepts plain CLR values (boxed as
// object?) and maps them to wire types; on decode it returns plain CLR values.

using System;

namespace PrismDb
{
    /// <summary>Record-format type tags.</summary>
    public static class Tag
    {
        public const int Null = 0x00;
        public const int Bool = 0x01;
        public const int Int32 = 0x02;
        public const int Int64 = 0x03;
        public const int Double = 0x04;
        public const int String = 0x05;
        public const int Binary = 0x06;
        public const int Timestamp = 0x09;
        public const int ObjectId = 0x0A;
    }

    /// <summary>A 12-byte document identifier.</summary>
    public sealed class ObjectId : IEquatable<ObjectId>
    {
        public byte[] Bytes { get; }

        public ObjectId(byte[] bytes)
        {
            if (bytes.Length != 12) throw new ProtocolException("ObjectId must be 12 bytes");
            Bytes = (byte[])bytes.Clone();
        }

        /// <summary>Lowercase 24-character hex.</summary>
        public string ToHex()
        {
            var c = new char[24];
            const string hex = "0123456789abcdef";
            for (int i = 0; i < 12; i++)
            {
                c[i * 2] = hex[Bytes[i] >> 4];
                c[i * 2 + 1] = hex[Bytes[i] & 0xF];
            }
            return new string(c);
        }

        public static ObjectId FromHex(string hex)
        {
            if (hex.Length != 24) throw new ProtocolException("ObjectId hex must be 24 chars");
            var b = new byte[12];
            for (int i = 0; i < 12; i++)
                b[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
            return new ObjectId(b);
        }

        public override string ToString() => ToHex();

        public bool Equals(ObjectId? other)
        {
            if (other is null || other.Bytes.Length != Bytes.Length) return false;
            for (int i = 0; i < Bytes.Length; i++)
                if (Bytes[i] != other.Bytes[i]) return false;
            return true;
        }

        public override bool Equals(object? obj) => Equals(obj as ObjectId);

        public override int GetHashCode()
        {
            int h = 17;
            foreach (var x in Bytes) h = h * 31 + x;
            return h;
        }
    }

    /// <summary>An explicitly-typed value, for cases where the default mapping of a
    /// CLR value is not what you want (e.g. a 32-bit int, a float that happens to
    /// be integral, or a timestamp). Build with the <see cref="Prism"/> helpers.</summary>
    public sealed class Typed
    {
        public int TypeTag { get; }
        public object Value { get; }

        public Typed(int tag, object value)
        {
            TypeTag = tag;
            Value = value;
        }
    }

    /// <summary>Helpers to build explicitly-typed values.</summary>
    public static class Prism
    {
        /// <summary>Force a value to wire Int32.</summary>
        public static Typed Int32(int n) => new Typed(Tag.Int32, n);
        /// <summary>Force a value to wire Int64.</summary>
        public static Typed Int64(long n) => new Typed(Tag.Int64, n);
        /// <summary>Force a value to wire Double.</summary>
        public static Typed Float64(double n) => new Typed(Tag.Double, n);
        /// <summary>Force a value to wire Timestamp (microseconds since the Unix epoch).</summary>
        public static Typed Timestamp(long us) => new Typed(Tag.Timestamp, us);
    }

    internal static class ValueCodec
    {
        private static readonly DateTime Epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        /// <summary>Resolve a CLR value to its wire type tag. Integer types map to
        /// Int64 to match the reference SDK; use Prism.Int32 to force Int32.</summary>
        public static int TagOf(object? v)
        {
            switch (v)
            {
                case null: return Tag.Null;
                case Typed t: return t.TypeTag;
                case ObjectId: return Tag.ObjectId;
                case bool: return Tag.Bool;
                case sbyte:
                case byte:
                case short:
                case ushort:
                case int:
                case uint:
                case long:
                case ulong: return Tag.Int64;
                case float:
                case double: return Tag.Double;
                case string: return Tag.String;
                case DateTime:
                case DateTimeOffset: return Tag.Timestamp;
                case byte[]: return Tag.Binary;
            }
            throw new ProtocolException($"unsupported value: {v.GetType().Name}");
        }

        private static long ToLong(object v) => Convert.ToInt64(v);
        private static double ToDouble(object v) => Convert.ToDouble(v);

        private static long EpochMicros(object v)
        {
            DateTimeOffset dto;
            if (v is DateTimeOffset d)
            {
                dto = d;
            }
            else
            {
                var dt = (DateTime)v;
                if (dt.Kind == DateTimeKind.Unspecified) dt = DateTime.SpecifyKind(dt, DateTimeKind.Utc);
                dto = new DateTimeOffset(dt);
            }
            return (dto.UtcDateTime - Epoch).Ticks / 10; // 100ns ticks -> microseconds
        }

        public static void EncodeUntagged(Writer w, int tag, object? v)
        {
            object? raw = v is Typed t ? t.Value : v;
            switch (tag)
            {
                case Tag.Null: break;
                case Tag.Bool: w.U8((bool)raw! ? 1 : 0); break;
                case Tag.Int32: w.I32((int)ToLong(raw!)); break;
                case Tag.Int64: w.I64(ToLong(raw!)); break;
                case Tag.Double: w.F64(ToDouble(raw!)); break;
                case Tag.Timestamp:
                    w.I64(raw is DateTime || raw is DateTimeOffset ? EpochMicros(raw!) : ToLong(raw!));
                    break;
                case Tag.String: w.StrU32((string)raw!); break;
                case Tag.ObjectId: w.Raw(((ObjectId)raw!).Bytes); break;
                case Tag.Binary:
                    var b = (byte[])raw!;
                    w.U32(b.Length);
                    w.U8(0); // subtype
                    w.Raw(b);
                    break;
                default:
                    throw new ProtocolException($"cannot encode value tag 0x{tag:x}");
            }
        }

        public static void EncodeTagged(Writer w, object? v)
        {
            int tag = TagOf(v);
            w.U8(tag);
            EncodeUntagged(w, tag, v);
        }

        public static object? DecodeUntagged(Reader r, int tag)
        {
            switch (tag)
            {
                case Tag.Null: return null;
                case Tag.Bool: return r.U8() != 0;
                case Tag.Int32: return r.I32();
                case Tag.Int64: return r.I64();
                case Tag.Double: return r.F64();
                case Tag.Timestamp: return r.I64();
                case Tag.String: return r.StrU32();
                case Tag.ObjectId: return new ObjectId(r.Raw(12));
                case Tag.Binary:
                    int len = (int)r.U32();
                    r.U8(); // subtype (discarded)
                    return r.Raw(len);
                default:
                    throw new ProtocolException($"unknown value tag 0x{tag:x}");
            }
        }

        public static object? DecodeTagged(Reader r) => DecodeUntagged(r, r.U8());
    }
}

File: sdks/python/prismdb/value.py
"""The scalar value model and its tagged/untagged wire codec.

Mirrors ``crates/prism-protocol/src/data.rs`` (``Value``) and the type tags in
``docs/specs/record-format.md``. The SDK accepts plain Python values and maps
them to wire types; on decode it returns plain Python values."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Union

from ._codec import Reader, Writer
from .errors import ProtocolError


class TAG:
    """Record-format type tags."""

    NULL = 0x00
    BOOL = 0x01
    INT32 = 0x02
    INT64 = 0x03
    DOUBLE = 0x04
    STRING = 0x05
    BINARY = 0x06
    TIMESTAMP = 0x09
    OBJECTID = 0x0A


class ObjectId:
    """A 12-byte document identifier."""

    __slots__ = ("bytes",)

    def __init__(self, raw: bytes) -> None:
        if len(raw) != 12:
            raise ProtocolError("ObjectId must be 12 bytes")
        self.bytes = bytes(raw)

    def to_hex(self) -> str:
        """Lowercase 24-character hex."""
        return self.bytes.hex()

    @classmethod
    def from_hex(cls, hex_str: str) -> "ObjectId":
        if len(hex_str) != 24:
            raise ProtocolError("ObjectId hex must be 24 chars")
        return cls(bytes.fromhex(hex_str))

    def __str__(self) -> str:
        return self.to_hex()

    def __repr__(self) -> str:
        return f"ObjectId('{self.to_hex()}')"

    def __eq__(self, other: object) -> bool:
        return isinstance(other, ObjectId) and other.bytes == self.bytes

    def __hash__(self) -> int:
        return hash(self.bytes)


class Typed:
    """An explicitly-typed value, for cases where the default mapping of a
    Python value is not what you want (e.g. a 32-bit int, a float that happens
    to be integral, or a timestamp). Build with :func:`int32` / :func:`int64` /
    :func:`float64` / :func:`timestamp`."""

    __slots__ = ("tag", "value")

    def __init__(self, tag: int, value: Union[int, float]) -> None:
        self.tag = tag
        self.value = value


def int32(n: int) -> Typed:
    """Force a value to wire ``Int32``."""
    return Typed(TAG.INT32, int(n))


def int64(n: int) -> Typed:
    """Force a value to wire ``Int64``."""
    return Typed(TAG.INT64, int(n))


def float64(n: float) -> Typed:
    """Force a value to wire ``Double``."""
    return Typed(TAG.DOUBLE, float(n))


def timestamp(us: int) -> Typed:
    """Force a value to wire ``Timestamp`` (microseconds since the Unix epoch)."""
    return Typed(TAG.TIMESTAMP, int(us))


# A value the SDK can send. Plain Python values map as: ``None``->Null,
# ``bool``->Bool, ``int``->Int64, ``float``->Double, ``str``->Str,
# ``bytes``->Binary, ``datetime``->Timestamp, ``ObjectId``->ObjectId. Wrap with
# int32/float64/timestamp for other types.
Value = Union[
    None, bool, int, float, str, bytes, bytearray, datetime, ObjectId, Typed
]


def tag_of(v: Value) -> int:
    """Resolve a Python value to its wire type tag."""
    if v is None:
        return TAG.NULL
    if isinstance(v, Typed):
        return v.tag
    if isinstance(v, ObjectId):
        return TAG.OBJECTID
    if isinstance(v, bool):  # must precede int - bool is a subclass of int
        return TAG.BOOL
    if isinstance(v, int):
        return TAG.INT64
    if isinstance(v, float):
        return TAG.DOUBLE
    if isinstance(v, str):
        return TAG.STRING
    if isinstance(v, datetime):
        return TAG.TIMESTAMP
    if isinstance(v, (bytes, bytearray)):
        return TAG.BINARY
    raise ProtocolError(f"unsupported value: {v!r}")


def _epoch_micros(dt: datetime) -> int:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return int(dt.timestamp() * 1_000_000)


def encode_untagged(w: Writer, tag: int, v: Value) -> None:
    """Write the value bytes for a known ``tag`` (no tag byte)."""
    raw = v.value if isinstance(v, Typed) else v
    if tag == TAG.NULL:
        return
    if tag == TAG.BOOL:
        w.u8(1 if raw else 0)
    elif tag == TAG.INT32:
        w.i32(int(raw))
    elif tag == TAG.INT64:
        w.i64(int(raw))
    elif tag == TAG.DOUBLE:
        w.f64(float(raw))
    elif tag == TAG.TIMESTAMP:
        w.i64(_epoch_micros(raw) if isinstance(raw, datetime) else int(raw))
    elif tag == TAG.STRING:
        w.str_u32(str(raw))
    elif tag == TAG.OBJECTID:
        w.raw(raw.bytes)  # type: ignore[union-attr]
    elif tag == TAG.BINARY:
        b = bytes(raw)  # type: ignore[arg-type]
        w.u32(len(b))
        w.u8(0)  # subtype
        w.raw(b)
    else:
        raise ProtocolError(f"cannot encode value tag 0x{tag:x}")


def encode_tagged(w: Writer, v: Value) -> None:
    """Write a tagged value (tag byte then the value bytes)."""
    tag = tag_of(v)
    w.u8(tag)
    encode_untagged(w, tag, v)


def decode_untagged(r: Reader, tag: int) -> Value:
    """Read the value bytes for a known ``tag``, returning a plain Python value."""
    if tag == TAG.NULL:
        return None
    if tag == TAG.BOOL:
        return r.u8() != 0
    if tag == TAG.INT32:
        return r.i32()
    if tag == TAG.INT64:
        return r.i64()
    if tag == TAG.DOUBLE:
        return r.f64()
    if tag == TAG.TIMESTAMP:
        return r.i64()
    if tag == TAG.STRING:
        return r.str_u32()
    if tag == TAG.OBJECTID:
        return ObjectId(r.raw(12))
    if tag == TAG.BINARY:
        n = r.u32()
        r.u8()  # subtype (discarded)
        return r.raw(n)
    raise ProtocolError(f"unknown value tag 0x{tag:x}")


def decode_tagged(r: Reader) -> Value:
    """Read a tagged value."""
    return decode_untagged(r, r.u8())
